Reporting/M4 Document/Testing Strategy.md

```diff
@@ -0,0 +1,12 @@
+1. # <a name="_4w5ju7g24iuv"></a>**Testing Strategy and Updates**
+### <a name="_6qln2i1louzi"></a>Current Test Approaches
+- Similar to what has been outlined in M3, we are sticking with 3 main testing approaches: ad-hoc, unit, and black box testing.
+  - For **Ad-hoc** testing: We use this approach mainly to check if the application is showing or behaving a certain way that we expect it to be. These tests are discretionary and done quickly.
+  - For **Unit** testing: We plan and develop them for features that we aim to implement within a given sprint before the features are implemented themselves.  In contrast to ad hoc tests, they are formally documented as test cases. For example, such tests would cover parsing JSON data from MongoDB queries and checking if the outputs are as expected. In addition, this may also involve checking if a query parameter always returns a unique object.
```

Comment:
I like that you mentioned how a lot of our testing involves parsing JSON data from MongoDB queries and how we need to account for whether a query parameter always returns a unique object. This was actually something that I realized while writing the backend API endpoints for our core features, as for example, the provider allows users to have the same username. To counter this, I used the user ID that is internally used by Clerk (our authentication provider) to guarantee that `findOne` queries always return exactly one document. I will need to modify the messaging feature to rely on the same internal user ID instead of usernames to prevent users from messaging themselves if the recipient has the same username as the sender. I have not tested whether the provider allows users with the same username, but I believe this is the case.